dataset/src/main/scala/frameless/TypedDataset.scala

```diff
@@ -1036,8 +1036,15 @@ object TypedDataset {
     val shouldReshape = output.zip(targetColNames).exists {
       case (expr, colName) => expr.name != colName
     }
-
-    val reshaped = if (shouldReshape) df.toDF(targetColNames: _*) else df
+    val canSelect = targetColNames.toSet.subsetOf(output.map(_.name).toSet)
+
+    val reshaped = if (shouldReshape && canSelect) {
+      df.select(targetColNames.head, targetColNames.tail:_*)
```

Comment:
It's not totally clear whether or not systematically doing a select would create a lot of additional work at runtime. What do you think?

Reply:
The thing is i don't know how to otherwise reshape the dataframe into the correct form. I would suspect this is a rather cheap operation as it only shifts references, as it is guaranteed to only fetch a subset of the existing fields.

Actually the plan for the test case `collect()` is:

```
Project [_1#2 AS b#11, _2#3 AS a#12]
+- LocalRelation [_1#2, _2#3]
```

which is just shifting the fields. 

And the select is only done if the column names are not already in the right order.